gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java

```diff
@@ -368,29 +369,40 @@ private void flush(String dbName, String tableName) throws IOException {
    * Note that this is one of the place where the materialization of aggregated metadata happens.
    * When there's a change of {@link OperationType}, it also interrupts metadata aggregation,
    * and triggers materialization of metadata.
+   *
+   * @param emitAllErrors if true, go through all datasets in {@link #datasetErrorMap} and emit GTE for each. We only want
+   *                      to do this if a watermark is is being advanced.
+   *
    * @throws IOException
    */
-  @Override
-  public void flush() throws IOException {
+  public void flush(boolean emitAllErrors) throws IOException {
     log.info(String.format("start to flushing %s records", String.valueOf(recordCount.get())));
     for (String tableString : tableOperationTypeMap.keySet()) {
       List<String> tid = Splitter.on(TABLE_NAME_DELIMITER).splitToList(tableString);
       flush(tid.get(0), tid.get(1));
     }
     tableOperationTypeMap.clear();
     recordCount.lazySet(0L);
-    // Emit events for all current errors, since the GMCE watermark will be advanced
-    for (Map.Entry<String, Map<String, GobblinMetadataException>> entry : datasetErrorMap.entrySet()) {
-      for (GobblinMetadataException exception : entry.getValue().values()) {
-        submitFailureEvent(exception);
+    if (emitAllErrors) {
```

Comment:
Should we just clear the datasetErrorMap when throw exception? Or do we actually need to flush() when writer is close? I think we only want flush when watermark is moving and this does not happen when close is called?

Reply:
My thought was that this way is better, because when we call flush(), there may be iceberg watermarks advanced for tables that actually do flush, and for those ones, we do want to send failure GTE (but not for ones that just failed and are not having a successful flush).

>I think we only want flush when watermark is moving
Did you mean to say only send GTE when watermark is moving or are you suggesting we remove the flush from close method?

Reply:
As discussed removed the flush from close() instead.